## Setting Up The Environment

In [5]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

In [6]:
#Create a path
path = r'C:\Users\manny\Documents\Instacart Basket Analysis 05-01-2023'

In [7]:
#import products.csv
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [8]:
#import orders_wrangled.csv
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## Mixed Type Data

In [9]:
#Create a datafrome

df_test=pd.DataFrame()

In [10]:
# Create a mixed type column
df_test['mix']=['a','b',1,True]

In [11]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [12]:
# Check for mixed data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [13]:
#Mix column is mostly strings.
df_test['mix'] = df_test['mix'].astype('str')

## Missing values

In [14]:
#Find missing value
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
#Create null subset
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
df_prods.shape

(49693, 5)

In [19]:
#Create df_prods_clean susbset
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
df_prods_clean.shape

(49677, 5)

## Duplicates


In [21]:
#Look for full duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [23]:
df_prods_clean.shape

(49677, 5)

In [24]:
#Drop dupes
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

In [27]:
# Export products csv as prepped data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

## Task

## 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [28]:
# Descriptive analysis on df_ords
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

## 3. Check for mixed-type data in your df_ords dataframe.

In [35]:
# Check for mixed data types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, ' mixed')
  else: print (col, ' single')

Unnamed: 0  single
order_id  single
user_id  single
order_number  single
orders_day_of_week  single
order_hour_of_day  single
days_since_prior_order  single


## 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

No mixed data

## 5. Run a check for missing values in your df_ords dataframe.

In [36]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [39]:
#View null values
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [44]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [45]:
# Min and Max of order number
df_ords_nan.describe()

Unnamed: 0      order_id        user_id  order_number  \
count  2.062090e+05  2.062090e+05  206209.000000      206209.0   
mean   1.712637e+06  1.708462e+06  103105.000000           1.0   
std    9.874816e+05  9.881299e+05   59527.555167           0.0   
min    0.000000e+00  2.000000e+01       1.000000           1.0   
25%    8.579730e+05  8.507300e+05   51553.000000           1.0   
50%    1.717068e+06  1.706246e+06  103105.000000           1.0   
75%    2.570518e+06  2.564292e+06  154657.000000           1.0   
max    3.421069e+06  3.421081e+06  206209.000000           1.0   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count       206209.000000      206209.000000                     0.0  
mean             2.754118          13.626597                     NaN  
std              2.076205           4.223769                     NaN  
min              0.000000           0.000000                     NaN  
25%              1.000000          11.000000                     NaN  
50%              3.000000          14.000000                     NaN  
75%              5.000000          17.000000                     NaN  
max              6.000000          23.000000                     NaN

The column 'days_since_prior_order' contains 206209 rows of null values.
A susbet was created to find all missing values.
When viewing the subset, it is noted that the head and tail of order number column is 1.
When the min and max of order number is checked, it is 1 which means all 206209 rows have a value of 1.

This makes sense that the column 'days_since_prior_order' would be null because this is the first order which means there would be no data of their prior order.

## 6. Address the missing values using an appropriate method

No action done. As mentioned, there simply is no data because this is the first order of the customer.

## 7. Run a check for duplicate values in your df_ords data

In [46]:
# Run duplicates on all column
df_ords_dups=df_ords[df_ords.duplicated()]

In [47]:
df_ords_dups.shape

(0, 7)

In 7 columns, there are 0 rows, hence no dupes found.

## 8. Address the duplicates using an appropriate method

No action needed, no dupes found.

In [49]:
# Export order csv as prepped data
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))